File: services/eval/src/rag_eval/rerank/cohere.py
from collections.abc import Sequence
from typing import Self

import httpx

from rag_eval.http_retry import post_with_retry
from rag_eval.rerank.base import RerankCandidate, RerankedResult


class CohereReranker:
    provider = "cohere"

    def __init__(
        self,
        api_key: str,
        model: str,
        base_url: str = "https://api.cohere.com/v2",
        client: httpx.Client | None = None,
    ) -> None:
        self._api_key = api_key
        self.model = model
        self._base_url = base_url
        self._client = client or httpx.Client(timeout=30.0)

    def rerank(
        self, query: str, candidates: Sequence[RerankCandidate], top_n: int
    ) -> list[RerankedResult]:
        if not candidates:
            return []
        data = post_with_retry(
            self._client,
            f"{self._base_url}/rerank",
            service="Cohere rerank",
            json={
                "model": self.model,
                "query": query,
                "documents": [content for _, content in candidates],
                "top_n": top_n,
            },
            headers={"Authorization": f"Bearer {self._api_key}"},
        )
        return [
            (candidates[result["index"]][0], float(result["relevance_score"]))
            for result in data["results"]
        ]

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> Self:
        return self

    def __exit__(self, *exc_info: object) -> None:
        self.close()

File: services/agent/src/rerank/cohere.ts
import type { RerankCandidate, RerankedResult, Reranker } from "./reranker.js";

const DEFAULT_BASE_URL = "https://api.cohere.com/v2";

interface CohereRerankResponse {
  results: Array<{ index: number; relevance_score: number }>;
}

export interface CohereRerankerOptions {
  readonly apiKey: string;
  readonly model: string;
  readonly baseUrl?: string;
  readonly fetchImpl?: typeof fetch;
}

/** Mirrors services/eval/src/rag_eval/rerank/cohere.py (POST /v2/rerank). */
export function createCohereReranker(options: CohereRerankerOptions): Reranker {
  const fetchImpl = options.fetchImpl ?? fetch;
  const baseUrl = options.baseUrl ?? DEFAULT_BASE_URL;

  return {
    provider: "cohere",
    model: options.model,
    async rerank(
      query: string,
      candidates: readonly RerankCandidate[],
      topN: number,
    ): Promise<RerankedResult[]> {
      if (candidates.length === 0) return [];
      const response = await fetchImpl(`${baseUrl}/rerank`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${options.apiKey}`,
        },
        body: JSON.stringify({
          model: options.model,
          query,
          documents: candidates.map((candidate) => candidate.content),
          top_n: topN,
        }),
      });
      if (!response.ok) {
        throw new Error(`Cohere rerank failed: ${response.status}`);
      }
      const data = (await response.json()) as CohereRerankResponse;
      return data.results.map((result) => ({
        id: candidates[result.index].id,
        score: result.relevance_score,
      }));
    },
  };
}
